internal/service/users.go

```diff
@@ -40,6 +45,49 @@ func (u *usersService) UpdateUserInfo(ctx context.Context, id string, input *Upd
 	return u.repo.Update(ctx, id, &upd)
 }
 
+func (u *usersService) Signup(ctx context.Context, input *SignupUserInput) error {
+	user, err := u.repo.GetByEmail(ctx, input.Email)
+	if err != nil && err != repository.ErrNotFound {
+		return err
+	}
+
+	if user != nil && user.Email == input.Email {
+		return ErrUserAlreadyExist
+	}
+
+	hashPassword, err := bcrypt.GenerateFromPassword([]byte(input.Password), bcrypt.MinCost)
+	if err != nil {
+		return err
+	}
+	user = &core.User{
+		Id:               u.idGen.Generate(),
+		Email:            input.Email,
+		FirstName:        input.FirstName,
+		LastName:         input.LastName,
+		DisplayName:      input.DisplayName,
+		RegistrationDate: time.Now(),
+		HashedPassword:   hashPassword,
+		Roles:            []security.Role{security.Student},
+	}
+
+	if err := u.repo.Insert(ctx, user); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (u *usersService) Login(ctx context.Context, input *LoginInput) (*core.User, error) {
+	user, err := u.repo.GetByEmail(ctx, input.Email)
+	if err != nil {
```

Comment:
This approach, combined with what you have in the endpoint implementation, will return different errors in the following cases:

- both email and password are incorrect
- correct email, incorrect password

A normal approach when building login functionality is to reply with the same error in both cases. This is also what we have in the API definition on SwaggerHub: error with status code 400 (bad request) and text "Email or password are incorrect"

Reply:
Updated in the latest commit.